airflow/utils/log/file_task_handler.py

```diff
@@ -0,0 +1,187 @@
+            log += "*** Fetching here: {url}\n".format(**locals())
+            try:
+                import requests
+                timeout = None  # No timeout
+                try:
+                    timeout = conf.getint('webserver', 'log_fetch_timeout_sec')
+                except (AirflowConfigException, ValueError):
+                    pass
+
+                response = requests.get(url, timeout=timeout)
+                response.raise_for_status()
+                log += '\n' + response.text
+            except Exception:
+                log += "*** Failed to fetch log file from worker.\n"
+
+        return log
+
+    def read(self, task_instance, try_number=None):
+        """
+        Read logs of given task instance from GCS remote storage. If failed,
+        read the log from task instance host machine.
+        :param task_instance: task instance object
+        :param try_number: task instance try_number to read logs from. If None
+                           it returns all logs separated by try_number
+        :return: a list of logs
+        """
+        current_try = task_instance.try_number
+
+        if try_number is None:
+            try_numbers = list(range(current_try))
+        elif try_number > current_try or try_number < 0:
+            logs = ['Error fetching the log. Try number {} is invalid.'.format(try_number)]
+            return logs
+        else:
+            try_numbers = [try_number]
+
+        logs = [''] * len(try_numbers)
+        for i, try_number in enumerate(try_numbers):
+            logs[i] += self._read(task_instance, try_number)
+
+        return logs
+
+    def _get_log_relative_dir(self, task_instance):
+        """
+        Get relative log file directory.
+        :param task_instance: task instance object
+        """
+        return os.path.join(task_instance.dag_id, task_instance.task_id,
+                            task_instance.execution_date.isoformat())
+
+    def _get_log_relative_path(self, task_instance, try_number=None):
+        """
+        Get relative log file path.
+        :param task_instance: task_instance object
+        :param try_number: optional task instance try_number to
+        override task instance try_number
+        """
+        directory = self._get_log_relative_dir(task_instance)
+        try_number = task_instance.try_number if try_number is None else try_number
+        # For reporting purposes and keeping logs consistent with web UI
+        # display, we write and report based on 1-indexed, not 0-indexed lists
+        filename = "{}.log".format(try_number + 1)
+        return os.path.join(directory, filename)
+
+    def _init_file(self, task_instance):
+        """
+        Create log directory and give it correct permissions.
+        :param task_instance: task instance object
+        :return relative log path of the given task instance
+        """
+        # To handle log writing when tasks are impersonated, the log files need to
```

Comment:
Also part of refactored code. This part, as well as this comment, should be changed in the future.